Updated markdown documentation files

StefanHabel opened the pull request:
Changes:
- Replaced relative image sources with absolute image URLs so that images load when viewed at https://pypi.org/project/MaterialX/ and when integrating the contents into the new Python API documentation
- Streamlined headline levels: page title (if any) uses `#`, first level uses `##`, second level uses `###`, etc.
- Used HTML tables for formatting image galleries
- Streamlined figure captions to format colons in **bold**
- Used HTML comment syntax for comments in XML blocks
- Other minor formatting changes, e.g. formatting the `MaterialXView` executable in a code font

Split from #1567.

Update #342.

StefanHabel commented:
Screenshot showing the new `Viewer.md` contents in the context of the work-in-progress Python API documentation:

![Screenshot 2024-09-29 at 17 51 16](https://github.com/user-attachments/assets/99d1f3c2-f125-4962-9de4-38b22dad6bfb)

jstone-lucasfilm commented:
Thanks for this proposal, @StefanHabel!

I think I'll need to spend some time with these new visual ideas to form a strong impression, though my initial sense is that the captions over the image rows on the front page (e.g. `Marble`, `Copper`, `Tiled Brass`, `Tiled Wood`) might be too distracting from the role these images play in the storyline of the front page.

Would it work visually if we adopted your new tabular organization of the image rows, but maintained the original placement of the `Figure N` text above each image row?

StefanHabel commented:
Thanks for the feedback @jstone-lucasfilm!

The different markdown pages currently use two different placements for the figure text:
- Above the figures: [`README.md`](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/README.md), [`GraphEditor.md`](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/DeveloperGuide/GraphEditor.md), [`Viewer.md`](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/DeveloperGuide/Viewer.md)
- Below the figures (including code snippets and tables): [`ShaderGeneration.md`](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/DeveloperGuide/ShaderGeneration.md)
    - https://materialx.org also uses a figure text below the image at the top

I've added two more commits to use HTML tables for these image galleries, with two different placements of figure texts -- above and below their corresponding figures -- so that we can compare the appearance.

In both cases, I've moved the descriptions of each image to small text below the respective image.

Above:

![Screenshot 2024-09-29 at 20 34 14](https://github.com/user-attachments/assets/739944b2-bfc1-4db0-84e9-27431751d208)

Below:

![Screenshot 2024-09-29 at 20 16 18](https://github.com/user-attachments/assets/2d32dcb4-484a-4d37-b303-4a63f12844ca)

Personally, I'm more used to text describing a figure appearing below the figure.

The current state in this branch is for the figure texts to appear above the figures, as before, but leaving those in `ShaderGeneration.md` as-is.

StefanHabel commented:
Maybe we could add a colon at the end of figure texts that appear above figures, to make it clearer that they pertain to the respective figure below?

e.g.

> **Figure 1:** Procedural and uniform materials in the MaterialX viewer:

```diff
-**Figure 1:** Procedural and uniform materials in the MaterialX viewer
+**Figure 1:** Procedural and uniform materials in the MaterialX viewer:
```

jstone-lucasfilm commented:
@StefanHabel Could it be that the individual labels for images (e.g. `Marble`, `Copper`) are not needed here, and that the storyline of the front page is clearer without them?

Just as a reference point, here's the current layout of the front page, which doesn't make use of these individual labels:

![image](https://github.com/user-attachments/assets/fd6ffa3c-86fd-43b3-acae-eab9ca8d9a6b)

StefanHabel commented:
Thanks Jonathan!

I've removed the small image captions from image galleries in https://github.com/AcademySoftwareFoundation/MaterialX/pull/2037/commits/dd275ae10234318ec64a9cb2245d71d1b36e46e3.

I've also restored two *Example Images* headlines on `Viewer.md` and `GraphEditor.md`.

Here's what the modified [`Viewer.md`](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/DeveloperGuide/Viewer.md) looks like in the context of the Python API documentation:

![Screenshot 2024-09-30 at 10 30 32](https://github.com/user-attachments/assets/acce7381-24e7-42c5-b98c-9b535aca1d6b)

jstone-lucasfilm commented:
The latest version of the front page looks good to me, thanks @StefanHabel!  I wonder if we can reduce the buffer widths in the image rows, minimizing the empty space between the images, but I don't believe that should be a blocking issue.

I'll start taking a closer look at the details of this change, and we can iterate from there.

StefanHabel commented:
Thanks Jonathan!

Here's what the image gallery tables would look like with `<table cellpadding="0" cellspacing="0">`:

![Screenshot 2024-09-30 at 15 00 27](https://github.com/user-attachments/assets/77290531-a3be-4149-aea4-dac121ad4f9c)

Is that what you had in mind?

jstone-lucasfilm commented:
Oh, that's more extreme than what I had in mind, and I was just picturing the original spacing between images from the pre-tabular version.

Here's the spacing that I currently see in your build:

![image](https://github.com/user-attachments/assets/a2e7b3a7-d80b-4bb5-b6f9-2e717c9fc907)

And here's the original spacing that I was hoping we could emulate:

![image](https://github.com/user-attachments/assets/dd29c86a-947d-4f1f-8a05-1fd8fc7f9573)

jstone-lucasfilm commented:
It's those large white blocks around each image that I was hoping we could remove, if there's an option in the HTML to allow this.

StefanHabel commented:
Understood!

We should be able to remove those by using `cellpadding="0"`, but leaving `cellspacing` to a small value.

We can also remove the table border using `border="0"`.

I'll experiment locally...

jstone-lucasfilm commented:
Much appreciated, @StefanHabel, and overall this looks like a great set of improvements!

I particularly enjoyed the update of a specification link that was still pointing to our v1.36 PDF instead of the GitHub Markdown for MaterialX v1.39.  :D

StefanHabel commented:
Thanks Jonathan!

Regarding linking to the MaterialX Specification, we currently use the following two different links (as of the current state of this PR):

1. `MainPage.md` links to https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/Specification/MaterialX.Specification.md
2. `ShaderGeneration.md` links to https://materialx.org/Specification.html

Should both links point to the same?

---

I've added https://github.com/AcademySoftwareFoundation/MaterialX/pull/2037/commits/ba35d10a14b3549312afa30262c2fbf0cf5d6f20 to tweak the image gallery spacing.

This is what the Viewer page now looks like in the context of the Python API docs:

![Screenshot 2024-09-30 at 15 29 55](https://github.com/user-attachments/assets/b7bf1ab2-7441-4b34-9074-8821c4c677da)

jstone-lucasfilm commented:
@StefanHabel That looks perfect in the screenshot above, though I get different results when I view the front page README of your fork of MaterialX:

https://github.com/StefanHabel/MaterialX/tree/%23342-markdown-changes

Let me know what you see on your side, and we can figure out what might account for that difference.

For linking to the specification, I think your choice to use https://materialx.org/Specification.html is the right one, as that covers the full set of specification documents as well as its history.

StefanHabel commented:
Argh, sorry, yeah, we can replicate this here, too, by pasting the HTML table in this comment:

<table style="border: none; border-collapse: separate; border-spacing: 4px;">
  <tbody>
    <tr>
      <td style="padding: 0px;">
        <img alt="A teapot with a Standard Surface Marble material applied"
             title="Standard Surface Marble material"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Marble.png">
      </td>
      <td style="padding: 0px;">
        <img alt="A teapot with a Standard Surface Copper material applied"
             title="Standard Surface Copper material"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Copper.png">
      </td>
      <td style="padding: 0px;">
        <img alt="A teapot with a Standard Surface Plastic material applied"
             title="Standard Surface Plastic material"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Plastic.png">
      </td>
      <td style="padding: 0px;">
        <img alt="A teapot with a Standard Surface Carpaint material applied"
             title="Standard Surface Carpaint material"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Carpaint.png">
      </td>
    </tr>
  </tbody>
</table>

Looks like GitHub does not respect the given `style="border: none; border-collapse: separate; border-spacing: 4px;"` instructions.

As we're not using image captions per image in these tables after all, perhaps we should revert this and use images in a paragraph element again (but with those full URLs to the image sources, and perhaps with a width set to `24%` instead of the hard-coded `204` pixels):

<p float="left">
  <img alt="A teapot with a Standard Surface Marble material applied"
             title="Standard Surface Marble material" width="24%"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Marble.png"/>
  <img alt="A teapot with a Standard Surface Copper material applied"
             title="Standard Surface Copper material" width="24%"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Copper.png"/>
  <img alt="A teapot with a Standard Surface Plastic material applied"
             title="Standard Surface Plastic material" width="24%"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Plastic.png"/>
  <img alt="A teapot with a Standard Surface Carpaint material applied"
             title="Standard Surface Carpaint material" width="24%"
             src="https://raw.githubusercontent.com/AcademySoftwareFoundation/MaterialX/main/documents/Images/MaterialXView_Carpaint.png"/>
</p>

StefanHabel commented:
Here's what the `Viewer.md` page looks like in the context of the Python API documentation now:

![Screenshot 2024-09-30 at 16 04 14](https://github.com/user-attachments/assets/febf003e-4264-447b-aeca-6daaf2e5f871)

The spacing between the four images at the top seems a little uneven, which may be due to rounding.

But it's perhaps closest to the original appearance still.

StefanHabel commented:
We're actually seeing this uneven spacing (possibly due to rounding errors) in the current version of `Viewer.md`, too, when opening the *Outline* on the right-hand side: comparing the current version with the proposed new version:
- https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/DeveloperGuide/Viewer.md
- https://github.com/StefanHabel/MaterialX/blob/%23342-markdown-changes/documents/DeveloperGuide/Viewer.md

jstone-lucasfilm commented:
Ah, that latest version looks quite good!  I think this achieves the goal of maintaining the appearance of the original image rows, using the more adaptable table-based implementation that you've proposed.

The slight variation in spacing doesn't seem problematic, and I think you're right that it was originally present in some viewing environments as well.

jstone-lucasfilm commented:
Oh, I've just caught up with the posts above, and I see that it's no longer "table based", but I think that your latest proposed change is a good compromise in any case.

StefanHabel commented:
oh, sorry, yeah, I've reverted the image galleries to `<p>` elements  again after all 😅

https://github.com/AcademySoftwareFoundation/MaterialX/pull/2037/commits/fe3a5fe6b7a6ac03f78eb8c4ed9bde7f85561292

I've also used the same link to the MaterialX Specification here:

https://github.com/AcademySoftwareFoundation/MaterialX/pull/2037/commits/b4ec7691d04732dab122cce142d7c2e27899dfa2

jstone-lucasfilm reviewed (approved):
This change looks good to me, thanks @StefanHabel!

StefanHabel commented:
Thanks for collaborating on this and merging it @jstone-lucasfilm! 🙌

Another positive side effect of the use of relative widths for the image galleries is that on mobile (and generally screens of varying sizes) the pages look much closer to the desktop experience:

| Before | After |
|-|-|
| ![IMG_9296](https://github.com/user-attachments/assets/4fdbd92d-7156-4d4c-b67c-3de29a36ba8b) | ![IMG_9295](https://github.com/user-attachments/assets/41ab4663-c21d-4155-8a48-299e6d2f421e) |

(Also noting the change of headline level.)

StefanHabel commented:
For the record/archive:

✅ This PR has resolved the following warnings that were previously issued by Sphinx/[MyST-Parser](https://pypi.org/project/myst-parser/) when integrating the markdown pages into the Python API documentation (see https://github.com/AcademySoftwareFoundation/MaterialX/pull/2038):

```python
GraphEditor.md.rst:5: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:5: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:12: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:22: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:57: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:61: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
MainPage.md.rst:82: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
ShaderGeneration.md:8: WARNING: image file not readable: documents/Images/shadergen.png [image.not_readable]
Viewer.md.rst:5: WARNING: Non-consecutive header level increase; H1 to H3 [myst.header]
GraphEditor.md:23: WARNING: 'myst' cross-reference target not found: '../../resources/Materials/Examples' [myst.xref_missing]
ShaderGeneration.md:4: WARNING: 'myst' cross-reference target not found: '/source/MaterialXGenGlsl' [myst.xref_missing]
ShaderGeneration.md:4: WARNING: 'myst' cross-reference target not found: '/source/MaterialXGenOsl' [myst.xref_missing]
ShaderGeneration.md:82: WARNING: 'myst' cross-reference target not found: '../Specification/MaterialX.v1.36.Spec.pdf' [myst.xref_missing]
ShaderGeneration.md:34: WARNING: Lexing literal_block '// Nodedef elements for node <add>\n<nodedef name="ND_add_float" node="add">\n  <input name="in1" type="float" />\n  <input name="in2" type="float" />\n  <output name="out" type="float" defaultinput="in1" />\n</nodedef>\n<nodedef name="ND_add_color3" node="add" type="color3">\n  <input name="in1" type="color3" />\n  <input name="in2" type="color3" />\n  <output name="out" type="color3" defaultinput="in1" />\n</nodedef>\n<... more types ...>\n\n// Implementation elements for node <add>\n<implementation name="IM_add_float" nodedef="ND_add_float" file="mx_add.inline" />\n<implementation name="IM_add_color3" nodedef="ND_add_color3" file="mx_add.inline" />\n<... more types ...>\n\n// Nodedef elements for node <mix>\n<nodedef name="ND_mix_float" node="mix">\n  <input name="fg" type="float" />\n  <input name="bg" type="float" />\n  <input name="mix" type="float" />\n  <output name="out" type="float" defaultinput="bg" />\n</nodedef>\n<nodedef name="ND_mix_color3" node="mix">\n  <input name="fg" type="color3" />\n  <input name="bg" type="color3" />\n  <input name="mix" type="color3" />\n  <output name="out" type="color3" defaultinput="bg" />\n</nodedef>\n<... more types ...>\n\n// Implementation elements for node <mix>\n<implementation name="IM_mix_float" nodedef="ND_mix_float" sourcecode="mix({{bg}}, {{fg}}, {{mix}})" />\n<implementation name="IM_mix_color3" nodedef="ND_mix_color3" sourcecode="mix({{bg}}, {{fg}}, {{mix}})" />\n<... more types ...>\n' as "xml" resulted in an error at token: 'm'. Retrying in relaxed mode. [misc.highlighting_failure]
Viewer.md:34: WARNING: 'myst' cross-reference target not found: '../../resources/Geometry' [myst.xref_missing]
Viewer.md:40: WARNING: 'myst' cross-reference target not found: '../../resources/Materials/Examples/StandardSurface' [myst.xref_missing]
Viewer.md:40: WARNING: 'myst' cross-reference target not found: '../../resources/Materials/Examples/UsdPreviewSurface' [myst.xref_missing]
Viewer.md:46: WARNING: 'myst' cross-reference target not found: '../../resources/Materials/Examples/StandardSurface/standard_surface_look_brass_tiled.mtlx' [myst.xref_missing]
Viewer.md:50: WARNING: 'myst' cross-reference target not found: '../../resources/Lights' [myst.xref_missing]
```
